torch/csrc/jit/passes/alias_analysis.h

```diff
@@ -49,6 +49,23 @@ class AliasDb {
   // hold in memory any element that exists in the other
   bool mayContainAlias(const Value* a, const Value* b) const;
 
+  // Do any values in group `a` share a memory location or hold in memory
+  // any element that exists in group `b`
+  template <
+      typename... Other1,
+      template <typename, typename...> class T,
+      typename... Other2,
+      template <typename, typename...> class U>
+  bool mayContainAlias(
```

Comment:
I feel like we keep re-introducing this inefficient pairwise comparison pattern in different levels of the code. Avoiding that is the reason `mayAlias` is implemented the way it is below, where we gather all memory locations before comparing them for overlap. Let's do the same here.

Reply:
I would rather avoid the way it's done below since it effectively implements mayAlias in two different locations. E.g. below this there is no check for 
`if (!shouldAnnotate(a) || !shouldAnnotate(b)) {` but that has since been added to the mayAlias implementation in the cpp file. 

If we do want to avoid gathering all memory locations, i think it makes the most sense to add an API in memory_dag.cpp for the group checking

Reply:
Good point, but that's a reason to delete the singleton variant of `mayAlias` I think (or at least make it a proxy for the container version). In general, we should be wary of code patterns that involve a lot of unnecessary work in passes/methods that are called really frequently.

I agree that we'll probably need to add to memory_dag (similar to how mayAlias has a memory_dag container comparison implementation as well).